src/main/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3.java

```diff
@@ -0,0 +1,101 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.nhs.hee.tis.trainee.forms.migration;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import io.mongock.api.annotations.ChangeUnit;
+import io.mongock.api.annotations.Execution;
+import io.mongock.api.annotations.RollbackExecution;
+import java.util.ArrayList;
+import java.util.List;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.core.env.Environment;
+import uk.nhs.hee.tis.trainee.forms.dto.enumeration.DeleteType;
+
+/**
+ * Add partial delete related metadata to existing forms on S3.
+ */
+@Slf4j
+@ChangeUnit(id = "AddPartialDeleteMetadataToS3", order = "7")
+public class AddPartialDeleteMetadataToS3 {
+
+  private final AmazonS3 amazonS3;
+  private final String bucketName;
+
+  private static final String FIXED_FIELDS =
+      "id,traineeTisId,lifecycleState,submissionDate,lastModifiedDate";
+
+  /**
+   * Add partial delete related metadata to existing forms on S3.
+   */
+  public AddPartialDeleteMetadataToS3(AmazonS3 amazonS3, Environment env) {
+    this.amazonS3 = amazonS3;
+    this.bucketName = env.getProperty("application.file-store.bucket");
+  }
+
+  /**
+   * Find all object on S3 and add metadata.
+   */
+  @Execution
+  public void migrate() {
+    log.info("Starting migration to add partial delete related metadata to existing forms on S3.");
+
+    try {
+      List<S3ObjectSummary> keyList = new ArrayList<>();
+      var objects = amazonS3.listObjects(bucketName);
+      keyList.addAll(objects.getObjectSummaries());
+
+      while (objects.isTruncated()) {
+        objects = amazonS3.listNextBatchOfObjects(objects);
+        keyList.addAll(objects.getObjectSummaries());
+      }
+      log.info("Updating {} objects in the bucket {}", keyList.size(), bucketName);
+
+      for (var obj : keyList) {
+        final var object = amazonS3.getObject(bucketName, obj.getKey());
+
+        var metadata = object.getObjectMetadata();
+        metadata.addUserMetadata("deletetype", DeleteType.PARTIAL.name());
+        metadata.addUserMetadata("fixedfields", FIXED_FIELDS);
+
+        final var request = new PutObjectRequest(
+            bucketName, obj.getKey(), object.getObjectContent(), metadata);
+        amazonS3.putObject(request);
```

Comment:
To make this more rerunnable (on failure) you should add a check here for whether the `deleteType`/`fixedFields` metadata already exists. If it does then we skip updating the object.

Reply:
Good shout. The `addUserMetadata()` would do an "upsert" to update if metadata exists/ insert if not exists. It should keep a correct result even if the migration is rerun though, it's good to skip the unneeded putObject operation if it's already updated.